src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,249 @@
+        mRootDirectorySupplier =
+                new DiskBasedAsyncCache.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+        mMaxCacheSizeInBytes = maxCacheSizeInBytes;
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final DiskBasedAsyncCache.CacheHeader entry = mEntries.get(key);
+        File file = getFileForKey(key);
+        Path path = Paths.get(file.getPath());
+        try {
+            AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate((int) file.length());
+            afc.read(
+                    buffer,
+                    0,
+                    buffer,
+                    new CompletionHandler<Integer, ByteBuffer>() {
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            if (attachment.hasArray()) {
+                                int offset = attachment.arrayOffset();
+                                byte[] data = attachment.array();
+                                cb.onGetComplete(entry.toCacheEntry(data));
+                            }
+                        }
+
+                        @Override
+                        public void failed(Throwable exc, ByteBuffer attachment) {
+                            cb.onGetComplete(null);
+                        }
+                    });
+        } catch (Exception e) {
+            cb.onGetComplete(null);
+        }
+    }
+
+    @Override
+    public void put(String key, Entry entry, OnPutCompleteCallback callback) {
+        // TODO (sphill99): Implement
+    }
+
+    /**
+     * Initializes the DiskBasedAsyncCache by scanning for all files currently in the specified root
+     * directory. Creates the root directory if necessary.
+     */
+    @Override
+    public synchronized void initialize() {
+        // TODO (sphill99): Implement
+    }
+
+    /**
+     * Invalidates an entry in the cache.
+     *
+     * @param key Cache key
+     * @param fullExpire True to fully expire the entry, false to soft expire
+     */
+    @Override
+    public synchronized void invalidate(String key, boolean fullExpire) {
+        // TODO (sphill99): Implement
+    }
+
+    /** Removes the specified key from the cache if it exists. */
+    @Override
+    public synchronized void remove(String key) {
+        // TODO (sphill99): Implement
+    }
+
+    /**
```

Comment:
I think some of these common functions that are the same between your new Cache implementations and the existing one could be factored out into a Util that all three classes could call

You may be able to do the same with some of the CONSTANTS at the top of the file that seem to be the same for all three

Reply:
Yeah, I want to extract some of the stuff that's repeated out to a higher level class. Should I do that for this PR?